src/lightning/fabric/strategies/xla_fsdp.py

```diff
@@ -0,0 +1,403 @@
+        """Given a folder, load the contents from a checkpoint and restore the state of the given objects.
+
+        The strategy currently only supports saving and loading sharded checkpoints which are stored in form of a
+        directory of multiple files rather than a single file.
+        """
+        if not _TORCH_GREATER_EQUAL_2_0:
+            raise NotImplementedError(
+                "Saving and loading checkpoints with the `FSDPStrategy` is not supported in PyTorch < 2.0."
+                " Please upgrade `torch` or file an issue: `https://github.com/Lightning-AI/lightning/issues`."
+            )
+        if not state:
+            raise ValueError(
+                f"Got XLAFSDPStrategy.load_checkpoint(..., state={state!r}) but a state with at least "
+                f" a model instance to reload is required. Pass it in like so:"
+                " FSDPStrategy.load_checkpoint(..., state={'model': model, ...})"
+            )
+
+        # broadcast the path from rank 0 to ensure all the states are loaded from a common path
+        path = self.broadcast(path)
+        if not os.path.isdir(path):
+            raise NotImplementedError(
+                f"The path `{path}` is a file or does not exist, but the `XLAFSDPStrategy` currently only supports loading from a"
+                " checkpoint(s) in a directory."
+            )
+        rank = self.local_rank
+        world_size = self.world_size
+
+        from torch_xla.distributed.fsdp import XlaFullyShardedDataParallel as XLAFSDP
+
+        modules = {key: module for key, module in state.items() if isinstance(module, XLAFSDP)}
+        optimizers = {key: optim for key, optim in state.items() if isinstance(optim, Optimizer)}
+        if self._state_dict_type == "sharded":
+            file = os.path.join(path, f"checkpoint_rank-{rank:08d}-of-{world_size:08d}.pth")
+            if not Path(file).is_file():
+                raise ValueError(
+                    f"The path {str(file)!r} does not point to valid sharded checkpoints. Make sure the path points to a"
+                    " directory with XLAFSDP checkpoint shards."
+                )
+            if len(modules) == 0:
+                raise ValueError(
+                    "Could not find a XLAFSDP model in the provided checkpoint state. Please provide the model as"
+                    " part of the state like so: `load_checkpoint(..., state={'model': model, ...})`. Make sure"
+                    " you set up the model (and optimizers if any) through the strategy before loading the checkpoint."
+                )
+            if len(modules) > 1:
+                raise ValueError(
+                    "Found multiple XLAFSDP modules in the given state. Loading checkpoints with FSDP is"
+                    " currently limited to a single model per checkpoint. To load multiple models, call the"
+                    " load method for each model separately with a different path."
+                )
```

Comment:
These two errors could be moved out of the sharded code path. The limitation applies to both sharded and full If i'm not mistaken

Reply:
Hm... when loading a full checkpoint, you do not need to check for 1 XLAFSDP model since the full checkpoint will not be FSDP wrapped. Look at https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist_fsdp_with_ckpt.py#L301-L303 as an example.